[Refactor] Eliminate circular imports | `3XX` -> `0`

flx-sta opened the pull request:
> [!CAUTION]
> PAY EXTRA ATTENTION TO `init-abilities`, `init-moves` etc. files!
> AS I HAVE TO MANUALLY ADOPT ANY CHANGES FROM THE ORIGINAL `all-*` FILES INTO THIS PR! 
  
## Why am I making these changes?

Eliminate the constant impact of circular dependenciesa

## What are the changes from a developer perspective?

* Please refer to the commit messages

> [!TIP]
> Added github action that executes and checks dependency injections

## How to test the changes?

- Game still starts & runs
- Tests succeed
- Run `npm run depcruise` to verify the given numbers

> [!TIP]
> `✔ no dependency violations found (1024 modules, 7619 dependencies cruised)`

## Checklist

- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
<!-- We have heavily optimized our test suite, so please actually run the tests :) -->
- [x] Are all unit tests still passing? (`npm run test:silent`)

----

> [!NOTE]
> PR description copied from #302

flx-sta committed:
add `is[Status|SelfStatus|Attack]Move()` to `Move`

667 dependency violations

flx-sta committed:
minor fix in battle-scene

667 dependency violations

flx-sta committed:
add `isSwitchPhase` to Phase

667 dependency violations

flx-sta committed:
add `isMovePhase` to Phase

667 dependency violations

flx-sta committed:
add `isSelectModifierPhase` to Phase

664 dependency violations

flx-sta committed:
add `isBerryModifier` to Modifier

664 dependency violations

flx-sta committed:
add `isPokemonHeldItemModifier` to Modifier

661 dependency violations

flx-sta committed:
add `isPersistentModifier` to Modifier

661 dependency violations

flx-sta committed:
add `isTerastallizeModifier` to Modifier

661 dependency violations

flx-sta committed:
add `isPokemonFormChangeItemModifier` to Modifier

660 dependency violations

flx-sta committed:
add `isSwitchEffectTransferModifier` to Modifier

658 dependency violations

flx-sta committed:
revert change on ModifierPredicate

flx-sta committed:
add `isCriticalCatchChanceBoosterModifier` to Modifier

649 dependency violations

flx-sta committed:
add `isHitHealModifier` to Modifier

649 dependency violations

flx-sta committed:
add `isPreserveBerryModifier` to Modifier

649 dependency violations

flx-sta committed:
add `isShinyRateBoosterModifier` to Modifier

649 dependency violations

flx-sta committed:
add `isPokemonMultiHitModifier` to Modifier

647 dependency violations

flx-sta committed:
Check for `Pokemon` via `.type`

646 dependency violations

flx-sta committed:
Check for `Trainer` via `.type`

646 dependency violations

flx-sta committed:
Move ME constants into constants.ts

439 dependency violations

flx-sta committed:
add `isLapsingPokemonHeldItemModifier` & `isLapsingPersistentModifier` to Modifier

437 dependency violations

flx-sta committed:
add `isAttackTypeBoosterModifier` to Modifier

437 dependency violations

flx-sta committed:
add `isAwaitableUiHandler` to UiHandler

437 dependency violations

flx-sta committed:
add `isTurnHealModifier` to Modifier

437 dependency violations

flx-sta committed:
add `isHealShopCostModifier` to Modifier

437 dependency violations

flx-sta committed:
add `is<Challenge>` to Challenge

441 dependency violations

flx-sta committed:
split up berry.ts

439 dependency violations

flx-sta committed:
put `selfStatLowerMoves` into own file

439 dependency violations

flx-sta committed:
put `PokemonMove` into own file

438 dependency violations

flx-sta committed:
split up `party-ui-handler`

444 dependency violations

flx-sta committed:
move "PartyUi.NoEffectMsg" to constants

445 dependency violations

flx-sta committed:
move "applyAbAttrs" into own file

382 dependency violations

flx-sta committed:
fix tests

- add missing Phaser import to event-bus
- Fix false `redirectPokemonMoves` change

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-02

flx-sta committed:
move `applyRevealedAbAttrs` into `apply-ab-attr.ts`

382 dependency violations

flx-sta committed:
Apply suggestions from code review

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
Apply suggestions from code review

382 dependency violations

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>
Co-authored-by: innerthunder <168692175+innerthunder@users.noreply.github.com>

flx-sta committed:
Check for `MysteryEncounterIntroVisuals` via `.type`

382 dependency violations

flx-sta committed:
Check for `BattleInfo` via `.type`

382 dependency violations

flx-sta committed:
fix check for PlayerBattleInfo

382 dependency violations

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-02

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-02

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-02

flx-sta committed:
Apply review suggestions/comments

383 dependency violations

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>
Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
remove self-stat-lower-moves array

and replace it with a function in `Move`

385 dependency violations

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
remove `t` import from constants

385 dependency violations

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
chore: clean up apply-ab-attrs

385 dependency violations

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
break apart `pokemon-species`

378 dependency violations

flx-sta committed:
Apply review suggestions/comments

378 dependency violations

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
remove `MistTag` from stat-stage-change-phase

378 dependency violations

flx-sta committed:
add `nextBattle()` and `gameOver` method to battlescene

380 dependency violations

flx-sta committed:
move `ForceSwitchOutHelper` to post-damage-force-switch-out-ab-attr

368 dependency violations

flx-sta committed:
chore: clean up import path in move.ts

368 dependency violations

flx-sta committed:
remove `AbstractConstructor<ArenaTag>` from arena.applyTags

363 dependency violations

flx-sta committed:
remove `AbstractConstructor<>` from arena.applyTagsForSide

362 dependency violations

flx-sta committed:
remove `AbstractConstructor<>` from pokemon.getTag

and add array as argument

344 dependency violations

flx-sta committed:
add `isMoveEffectPhase` to Phase.

339 dependency violations

flx-sta committed:
use `BattleScene.gameOver` in FaintPhase

336 dependency violations

flx-sta committed:
use `BattleScene.gameOver` in VictoryPhase

330 dependency violations

flx-sta committed:
use `BattleScene.gameOver` in MysteryEncounterBattleStartCleanupPhase

290 dependency violations

flx-sta committed:
exlcude vite plugins from depcruise

289 dependency violations

flx-sta committed:
move PartyUiHandler statics into utils

283 dependency violations

flx-sta committed:
add `isMysteryEncounterPostSummonTag()`

278 dependency violations

flx-sta committed:
move `initTrainerTypeDialogue` into own file

275 dependency violations

flx-sta committed:
add `EncoreAttr`

272 dependency violations

flx-sta committed:
add `isTypeBoostTag()`

271 dependency violations

flx-sta committed:
move `failOnGravityCondition` into own file

270 dependency violations

flx-sta committed:
introduce PokemonSpeciesForm.type

to avoid `instanceof` usage

267 dependency violations

flx-sta committed:
CounterDamageAttr replace `Move` with `Moves`

265 dependency violations

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-02

flx-sta committed:
Merge branch 'refactor/depcruise-02' into refactor/depcruise-03

265 dependency violations

flx-sta committed:
Move SpeciesFormChange trigger into own files

243 dependency violations

flx-sta committed:
merge importsin encounter-phase-utils

243 dependency violations

flx-sta committed:
move `PokemonSummonData` into own file

227 dependency violations

flx-sta committed:
organize imports in `all-moves.ts`

flx-sta committed:
add `faintBattler` method to BattleScene

200 dependency violations

flx-sta committed:
add `queuePokemonHeal` method to BattleScene

183 dependency violations

flx-sta committed:
allow `Moves/MoveId` as argument in MovePhase constructor

180 dependency violations

flx-sta committed:
patch ME constants names & docs

378 dependency violations

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
Merge branch 'refactor/depcruise-02' into refactor/depcruise-03

flx-sta committed:
add `MoveAttr.isCopyMoveAttr()`

178 dependency violations

flx-sta committed:
add 3 more `MoveAttr.is<SpecificAttr>()` methods

175 dependency violations

flx-sta committed:
move `applyChallenges` into challenge-utils

150 dependency violations

flx-sta committed:
introduce `AbAttrId` enum.

147 dependency violations

flx-sta committed:
move ability-utils into utils/

147 dependency violations

flx-sta committed:
move `getWeatherCondition` into ability-utils

147 dependency violations

flx-sta committed:
move `allAbilities` and create `init-abilites.ts`

146 dependency violations

flx-sta committed:
move `getAbApplyFunc` into apply-ab-attr.ts

and remove export from `applyRevealedAbAttrs`

146 dependency violations

flx-sta committed:
move `initMoveAnim` into own file

150 dependency violations

flx-sta committed:
move `initMoveChargeAnim` into own file

152 dependency violations

flx-sta committed:
move `loadMoveAnimAssets` into own file

158 dependency violations

flx-sta committed:
move `initCommonAnims` into own file

158 dependency violations

flx-sta committed:
move `AnimConfig` into own file

- move `AnimBlendType` into enum file
- move `AnimFocus` into enum file

158 dependency violations

flx-sta committed:
move `initEncounterAnims` into own file

158 dependency violations

flx-sta committed:
move `loadCommonAnimAssets` into `common-anim-utils.ts`

158 dependency violations

flx-sta committed:
move `loadEncounterAnimAssets` into `encounter-anim-utils.ts`

157 dependency violations

flx-sta committed:
move `populateAnims` into own file

157 dependency violations

flx-sta committed:
move anim maps into own files

154 dependency violations

flx-sta committed:
move `MoveChargeAnim` into own file

154 dependency violations

flx-sta committed:
move `CommonBattleAnim` into own file

154 dependency violations

flx-sta committed:
move `MoveAnim` into own file

151 dependency violations

flx-sta committed:
move `EncounterBattleAnim` into own file

151 dependency violations

flx-sta committed:
clean up `battle-anims`

151 dependency violations

flx-sta committed:
fix depcruise config ignoring ts files

39 dependency violations

flx-sta committed:
remove `SortDirection` enum (unused)

38 dependency violations

flx-sta committed:
remove `SessionHistoryResult` enum (unused)

37 dependency violations

flx-sta committed:
remove `debug.js` (unused)

36 dependency violations

flx-sta committed:
move encounter anims out of enum (whops)

35 dependency violations

flx-sta committed:
move `isLocal` and `isConnected` to api

from utils

27 dependency violations

flx-sta committed:
add `toTitleScreen` method to `BattleScene`

26 dependency violations

flx-sta committed:
make `Pokemon.getTag` use varargs

26 dependency violations

flx-sta committed:
Introduce `AbAttrFlag` to identify AbAttr

instead of using `instanceof`

34 dependency violations

flx-sta committed:
add `chargeMove()` to `BattleScene`

33 dependency violations

flx-sta committed:
add `toLoginScreen()` to `BattleScene`

32 dependency violations

flx-sta committed:
introduce `AchvCategory` enum

31 dependency violations

flx-sta committed:
extend `isGameOver` method

30 dependency violations

flx-sta committed:
add more modifier helper methods

29 dependency violations

flx-sta committed:
add `selectTarget` method to battle-scene

28 dependency violations

flx-sta committed:
add `isTurnHeldItemTransferModifier` to Modifier

and fix imports in achv.ts

27 dependency violations

flx-sta committed:
replace imports `#app/enums/...` -> `#enums/...`

27 dependency violations

flx-sta committed:
Split up move-conditions

27 dependency violations

flx-sta committed:
merge both `move-utils` (whops)

27 dependency violations

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
add `move` to `TurnMove`

(instead of just move-id)

26 dependency violations

flx-sta committed:
add `lastMove` to `Battle`

(instead of just lastMoveId)

25 dependency violations

flx-sta committed:
add `queueMoveChargeAnimation` to BattleScene

24 dependency violations

flx-sta committed:
chore: fix eslint

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

DayKev commented on src/battle-scene.ts:
```diff
@@ -201,6 +200,47 @@ export interface InfoToggle {
+
+interface ToTitleScreenInit {
+  /** Whether to add the {@linkcode TitlePhase} to the front of the phase queue or defer it. */
+  eager?: boolean;
```
These should both probably be the same, either both `eager` or both `defer` with the same `true`/`false` behavior.

DayKev commented on src/data/ab-attrs/post-damage-force-switch-ab-attr.ts:
```diff
@@ -0,0 +1,269 @@
+ * @extends PostDamageAbAttr
+ * @see {@linkcode applyPostDamage}
+ */
+
```
```suggestion
```

DayKev commented on src/data/ab-attrs/post-damage-force-switch-ab-attr.ts:
```diff
@@ -0,0 +1,269 @@
+ * @param pokemon - The Pokémon whose Shell Bell recovery is being calculated.
+ * @returns The amount of health recovered by Shell Bell.
+ */
+
```
```suggestion
```

DayKev commented on public/locales:
Reminder to remove locales commit.

flx-sta replied:
Nah. Idk how to strip that out easily. I've just updated it to the most recent `main` now

DayKev commented on src/data/ab-attrs/post-defend-ability-give-ab-attr.ts:
```diff
@@ -5,21 +5,22 @@ import { getPokemonNameWithAffix } from "#app/messages";
-      && !attacker.getAbility().hasAttr(UnsuppressableAbilityAbAttr)
-      && !attacker.getAbility().hasAttr(PostDefendAbilityGiveAbAttr)
+      && !attacker.getAbility().hasAttr(AbAttrFlag.UNSUPPRESSABLE_ABILITY)
+      && !attacker.getAbility().hasAttr(AbAttrFlag.POST_DEFEND_ABILITY_GIVE)
```
Note to self/everyone: should passives be treated differently than regular abilities? Discuss later.

flx-sta replied:
Is there an issue to track this? Otherwise it might be lost as soon as `Resolve Conversation` is hit

DayKev replied:
Made a discussions post #429

DayKev reviewed (commented):
Got through to the end of the `AbAttr`s so far.

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
update locales

flx-sta committed:
Apply review comments

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

MokaStitcher commented on src/constants.ts:
```diff
@@ -78,7 +78,7 @@ export const ME_AVERAGE_ENCOUNTERS_PER_RUN_TARGET = 12;
  * Will increase/decrease the chance of spawning a ME based on the current run's total MEs encountered vs {@linkcode ME_AVERAGE_ENCOUNTERS_PER_RUN_TARGET}
  * @example:
- * Average-Encounters-Per-Run = 17 (expects avg 1 ME every 10 floors)
+ * Average-Encounters-Per-Run = 12 (expects avg 1 ME every 10 floors)
```
```suggestion
 * Average-Encounters-Per-Run = 17 (expects avg 1 ME every 10 floors)
```
we'll get there 😆

flx-sta committed:
add `useMove` to BattleScene

20 dependency violations

MokaStitcher commented on src/battle.ts:
```diff
@@ -100,6 +101,7 @@ export default class Battle {
   public postBattleLoot: PokemonHeldItemModifier[] = [];
   public escapeAttempts: number = 0;
   public lastMoveId: MoveId;
+  public lastMove: Move;
```
hm that's confusing. Do we need both a lastMoveId and lastMove?
At the very least the Move has a reference to its id, so lastMoveId could be removed (unless lastMove is not always defined in which cased it should be reflected in its declaration)

flx-sta replied:
I simply added `lastMove` to get rid of more cross references and I try to keep the scope but I agree, I'll replace it

flx-sta committed:
fix `achv-category` filename

20 dependency violations

MokaStitcher commented on src/data/ab-attrs/ignore-move-effect-ab-attr.ts:
```diff
@@ -9,6 +10,11 @@ import { PreDefendAbAttr } from "./pre-defend-ab-attr";
  * @see {@linkcode applyPreDefend}
  */
 export class IgnoreMoveEffectsAbAttr extends PreDefendAbAttr {
```
huh, this (and the flag) is named IgnoreMoveEffect**s** but the file name doesn't have the 's'

MokaStitcher commented on src/data/ability.ts:
```diff
@@ -54,21 +41,21 @@ export class Ability implements Localizable {
    */
-  hasAttr<T extends AbAttr>(attrType: AbstractConstructor<T>): boolean {
-    return this.attrs.some((attr) => attr instanceof attrType);
+  hasAttr(flag: AbAttrFlag): boolean {
```
should this one be renamed `hasAttrFlag`?

flx-sta replied:
Yes

MokaStitcher commented on src/data/all-moves.ts:
```diff
@@ -458,7 +462,7 @@ export function initMoves() {
       .attr(WeightPowerAttr),
     new AttackMove(MoveId.COUNTER, ElementalType.FIGHTING, MoveCategory.PHYSICAL, -1, 100, 20, -1, -5, 1)
-      .attr(CounterDamageAttr, (move: Move) => move.category === MoveCategory.PHYSICAL, 2)
+      .attr(CounterDamageAttr, (move) => allMoves[move].category === MoveCategory.PHYSICAL, 2)
```
```suggestion
      .attr(CounterDamageAttr, (moveId) => allMoves[moveId].category === MoveCategory.PHYSICAL, 2)
```

MokaStitcher commented on src/data/all-moves.ts:
```diff
@@ -1602,7 +1603,8 @@ export function initMoves() {
         CounterDamageAttr,
-        (move: Move) => move.category === MoveCategory.PHYSICAL || move.category === MoveCategory.SPECIAL,
+        (moveId) =>
+          allMoves[moveId].category === MoveCategory.PHYSICAL || allMoves[moveId].category === MoveCategory.SPECIAL,
```
```suggestion
        (moveId) => allMoves[moveId].isAttackMove(),
```
I think based on your changes in the last circ dep PR this can replace the check?

MokaStitcher commented on src/data/all-moves.ts:
```diff
@@ -3960,7 +3962,8 @@ export function initMoves() {
         CounterDamageAttr,
-        (move: Move) => move.category === MoveCategory.PHYSICAL || move.category === MoveCategory.SPECIAL,
+        (moveId) =>
+          allMoves[moveId].category === MoveCategory.PHYSICAL || allMoves[moveId].category === MoveCategory.SPECIAL,
```
```suggestion
        (moveId) => allMoves[moveId].isAttackMove(),
```
same thing

MokaStitcher commented on src/data/move-attrs/copy-move-attr.ts:
```diff
@@ -29,14 +29,16 @@ export class CopyMoveAttr extends OverrideMoveEffectAttr {
-    globalScene.unshiftPhase(
-      new MovePhase(user as PlayerPokemon, targets, new PokemonMove(lastMove, 0, 0, true), true),
-    );
+    globalScene.unshiftPhase(new MovePhase(user as PlayerPokemon, targets, lastMove, true));
```
could we rename `lastMove` to `lastMoveId`?

MokaStitcher commented on src/data/move-attrs/instant-charge-attr.ts:
```diff
@@ -2,7 +2,7 @@ import type { Pokemon } from "#app/field/pokemon";
 import type { Move } from "#app/data/move";
 import { MoveAttr } from "#app/data/move-attrs/move-attr";
-import type { UserMoveConditionFunc } from "../move-conditions";
+import type { UserMoveConditionFunc } from "../../@types/UserMoveConditionFunc";
```
```suggestion
import type { UserMoveConditionFunc } from "#app/@types/UserMoveConditionFunc";
```

MokaStitcher commented on src/data/move-attrs/miss-effect-attr.ts:
```diff
@@ -1,7 +1,7 @@
 import type { Move } from "#app/data/move";
 import { MoveAttr } from "#app/data/move-attrs/move-attr";
-import type { UserMoveConditionFunc } from "../move-conditions";
+import type { UserMoveConditionFunc } from "../../@types/UserMoveConditionFunc";
```
```suggestion
import type { UserMoveConditionFunc } from "#app/@types/UserMoveConditionFunc";
```

MokaStitcher commented on src/data/move-attrs/moveset-copy-move-attr.ts:
```diff
@@ -21,7 +21,8 @@ export class MovesetCopyMoveAttr extends OverrideMoveEffectAttr {
     }
 
-    const copiedMove = allMoves[targetMoves[0].moveId];
+    // const copiedMove = allMoves[targetMoves[0].moveId];
```
```suggestion
```

MokaStitcher commented on src/data/move-attrs/no-effect-attr.ts:
```diff
@@ -1,7 +1,7 @@
 import type { Move } from "#app/data/move";
 import { MoveAttr } from "#app/data/move-attrs/move-attr";
-import type { UserMoveConditionFunc } from "../move-conditions";
+import type { UserMoveConditionFunc } from "../../@types/UserMoveConditionFunc";
```
```suggestion
import type { UserMoveConditionFunc } from "#app/@types/UserMoveConditionFunc";
```

MokaStitcher commented on src/data/move-attrs/encore-attr.ts:
```diff
@@ -0,0 +1,42 @@
+
+  override getCondition(): MoveConditionFunc | null {
+    return (_user, target, _move): boolean => {
+      if (target.isMax()) {
```
would it make sense to use the failOnMaxCondition instead of having that here? Or is it better to have all conditions in one place?

flx-sta replied:
Something to consider but I'd say to for now out of scope.

MokaStitcher commented on src/battle-scene.ts:
```diff
@@ -3628,4 +3668,103 @@ export default class BattleScene extends SceneBase {
+  }
+
+  gameOver({ isVictory, clearPhaseQueue }: GameOverInit = {}): void {
+    if (clearPhaseQueue) this.clearPhaseQueue();
```
```suggestion
    if (clearPhaseQueue) {
      this.clearPhaseQueue();
    }
```

flx-sta replied:
@MokaStitcher Where is `clearPhaseQueue` coming from? It's throwing an error if I add

MokaStitcher replied:
? The suggestion just changes your code from
`if (clearPhaseQueue) this.clearPhaseQueue();`
to 
```ts
    if (clearPhaseQueue) {
      this.clearPhaseQueue();
    }
```
for code consistency, since I don't think we've been using single line if statements

flx-sta replied:
oh. It was showing up somewhere completely different in my vs-code. Sorry (some Attr file)

MokaStitcher commented on src/data/weather.ts:
```diff
@@ -150,10 +151,10 @@ export class Weather {
       if (!suppressWeatherEffectAbAttr) {
         suppressWeatherEffectAbAttr = pokemon.hasPassive()
-          ? pokemon.getPassiveAbility().getAttrs(SuppressWeatherEffectAbAttr)[0]
+          ? pokemon.getPassiveAbility().getAttrs<SuppressWeatherEffectAbAttr>(AbAttrFlag.SUPPRESS_WEATHER_EFFECT)[0]
```
Not really the scope of this PR but couldn't `pokemon.getAbilityAttrs<SuppressWeatherEffectAbAttr>(AbAttrFlag.SUPPRESS_WEATHER_EFFECT)` be used instead of checking ability and passive like this?

(if yes can you maybe add a todo for it? because this seems like it needs more checks than this. For example if both passive and ability affect the weather the passive will be ignored, even if it's potentially better than the ability because if affects immutable weather for example)

flx-sta replied:
Waiting for more opinions on this ... (discord)

flx-sta replied:
This causes a stack-overflow in the tests

MokaStitcher commented on src/enums/achv-flag.ts:
```diff
@@ -0,0 +1,9 @@
+export enum AchvCategory {
```
can you rename the file to `achv-category` to match the enum?

flx-sta committed:
introduce `PhaseId`

18 dependency violations

MokaStitcher commented on src/enums/sort-direction.ts:
```diff
@@ -1,4 +0,0 @@
-export enum SortDirection {
```
huh. It seems like you extracted this enum as well as DropDownColumn, DropDownType, DropDownState and SortCriteria in #302 but didn't remove them from `dropdown.ts` 
SortDirection does not seem to be used outside of dropdown.ts so it's fine to remove this file, but could you remove the other enums from dropdown.ts? (and remove the export on SortDirection since it's not needed at the moment)

flx-sta replied:
fixed

MokaStitcher commented on src/field/pokemon-move.ts:
```diff
@@ -5,17 +5,17 @@ import { toDmgValue } from "#app/utils";
- * It links to {@linkcode Move} class via the move ID.
- * Compared to {@linkcode Move}, this class also tracks if a move has received.
+ * It links to {@linkcode MoveId} class via the move ID.
+ * Compared to {@linkcode MoveId}, this class also tracks if a move has received.
```
```suggestion
 * Compared to {@linkcode Move}, this class also tracks if a move has received
```
I think that's what was meant in the original comment

MokaStitcher commented on src/field/pokemon-move.ts:
```diff
@@ -5,17 +5,17 @@ import { toDmgValue } from "#app/utils";
  * PP Ups, amount of PP used, and things like that.
  * @see {@linkcode isUsable} - checks if move is restricted, out of PP, or not implemented.
- * @see {@linkcode getMove} - returns {@linkcode Move} object by looking it up via ID.
+ * @see {@linkcode getMove} - returns {@linkcode MoveId} object by looking it up via ID.
```
```suggestion
 * @see {@linkcode getMove} - returns {@linkcode Move} object by looking it up via ID.
```

MokaStitcher commented on src/field/pokemon-move.ts:
```diff
@@ -5,17 +5,17 @@ import { toDmgValue } from "#app/utils";
  * @see {@linkcode getMovePp} - returns amount of PP a move currently has.
  * @see {@linkcode getPpRatio} - returns the current PP amount / max PP amount.
- * @see {@linkcode getName} - returns name of {@linkcode Move}.
+ * @see {@linkcode getName} - returns name of {@linkcode MoveId}.
```
```suggestion
 * @see {@linkcode getName} - returns name of the {@linkcode Move}.
```

MokaStitcher commented on src/phases/faint-phase.ts:
```diff
@@ -189,7 +188,7 @@ export class FaintPhase extends PokemonPhase {
       const legalPlayerPartyPokemon = legalPlayerPokemon.filter((p) => !p.isActive(true));
       if (!legalPlayerPokemon.length) {
-        globalScene.unshiftPhase(new GameOverPhase());
+        globalScene.gameOver({ clearPhaseQueue: true });
```
```suggestion
        globalScene.gameOver({ clearPhaseQueue: false });
```
it doesn't seem like the phase queue was being cleared before here?

MokaStitcher commented on src/phases/mystery-encounter-phases/battle-start-cleanup-phase.ts:
```diff
@@ -62,7 +61,7 @@ export class MysteryEncounterBattleStartCleanupPhase extends Phase {
     const legalPlayerPartyPokemon = legalPlayerPokemon.filter((p) => !p.isActive(true));
     if (!legalPlayerPokemon.length) {
-      globalScene.unshiftPhase(new GameOverPhase());
+      globalScene.gameOver({ clearPhaseQueue: true });
```
```suggestion
      globalScene.gameOver({ clearPhaseQueue: false });
```
also here, clearPhaseQuere wasn't being called before?

MokaStitcher commented on src/phases/summon-phase.ts:
```diff
@@ -59,7 +58,7 @@ export class SummonPhase extends PartyMemberPokemonPhase {
         console.error("Party Details:\n", party);
         console.error("All available Pokemon were fainted or illegal!");
         globalScene.clearPhaseQueue();
```
```suggestion
```

MokaStitcher commented on src/ui/battle-info.ts:
```diff
@@ -17,7 +17,7 @@ import { WindowVariant } from "#enums/window-variant";
 import { ExpGainsSpeed } from "#enums/exp-gains-speed";
 import { settings } from "#app/system/settings/settings-manager";
-import { CANVAS_SCALE, GAME_WIDTH } from "#app/ui-constants";
+import { CANVAS_SCALE } from "#app/ui-constants";
```
```suggestion
import { CANVAS_SCALE, GAME_WIDTH } from "#app/ui-constants";
```
merge issue seems like

MokaStitcher commented on src/ui/battle-info.ts:
```diff
@@ -955,7 +955,7 @@ export default class BattleInfo extends Phaser.GameObjects.Container {
 export class PlayerBattleInfo extends BattleInfo {
   constructor() {
-    super(Math.floor(GAME_WIDTH) - 10, -72, true);
+    super(Math.floor(globalScene.game.canvas.width / 6) - 10, -72, true);
```
```suggestion
    super(GAME_WIDTH - 10, -72, true);
```

MokaStitcher commented on src/ui/challenges-select-ui-handler.ts:
```diff
@@ -365,7 +364,7 @@ export default class GameChallengesUiHandler extends UiHandler {
         this.updateChallengeArrows(this.startCursor.visible);
       } else {
         globalScene.clearPhaseQueue();
```
```suggestion
```

MokaStitcher reviewed (commented):
322/400 files

flx-sta committed:
eliminate `allMoves` from protect-attr & resist-last-move-type-attr

16 dependency violations

flx-sta committed:
Extend `useMove` in BattleScene

14 dependency violations

flx-sta committed:
Create `encounter-visuals-utils` file

14 dependency violations

flx-sta committed:
remove `allMoves` from targetMoveCopiableCondition

12 dependency violations

flx-sta committed:
move `applyMoveAttrsInternal` into move-utils

11 dependency violations

flx-sta committed:
chore: moveset-copy-move-attr import cleanup

11 dependency violations

flx-sta committed:
split up `move.ts` slightly

10 dependency violations

flx-sta committed:
fix tests and some other stuff

10 dependency violations

flx-sta committed:
apply review suggestions

10 dependency violations

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
apply review suggestions (2)

10 dependency violations

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
fix Weather.isEffectSuppressed

10 dependency violations

Co-authored-by: innerthunder <168692175+innerthunder@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
Revert "fix Weather.isEffectSuppressed"

This reverts commit 584e3056e6a6912618de06f27908272807a44ce0.

flx-sta committed:
restore `Weather.isEffectSuppressed`

PigeonBar commented:
You can get rid of all the `all-moves.ts` circ deps by moving the `initMoves()` function by itself to a new file (similarly to `init-species.ts`).

PigeonBar commented:
The `heal-status-effect-attr.ts` circ dep can be removed by moving this logic in `move.ts`:
https://github.com/Despair-Games/poketernity/blob/720b46bcf4c2b84857b61098fd583a7f6885b93b/src/data/move.ts#L891-L893
...over to this loop inside `initMoves()`:
https://github.com/Despair-Games/poketernity/blob/720b46bcf4c2b84857b61098fd583a7f6885b93b/src/data/all-moves.ts#L4068-L4071

flx-sta commented:
> You can get rid of all the `all-moves.ts` circ deps by moving the `initMoves()` function by itself to a new file (similarly to `init-species.ts`).

Now... I feel dumb. I should have thought of that, thank you :D

flx-sta committed:
move `init-moves` into own file

1 dependency violations

flx-sta committed:
add `initModifierTypes` and `modifierTypePools`

✔ no dependency violations found

flx-sta committed:
add depcruise workflow

flx-sta committed:
update dependency cruiser severities

all `error` so check fails properly

flx-sta committed:
make depdency-check workflow ignore submodule

We don't need it for that

flx-sta committed:
rename to depdendencies.yml

from `depdendency-check.yml`

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta commented:
> The `heal-status-effect-attr.ts` circ dep can be removed by moving this logic in `move.ts`:
> 
> https://github.com/Despair-Games/poketernity/blob/720b46bcf4c2b84857b61098fd583a7f6885b93b/src/data/move.ts#L891-L893
> 
> ...over to this loop inside `initMoves()`:
> 
> https://github.com/Despair-Games/poketernity/blob/720b46bcf4c2b84857b61098fd583a7f6885b93b/src/data/all-moves.ts#L4068-L4071

done

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

MokaStitcher commented on src/battle.ts:
```diff
@@ -99,7 +99,8 @@ export default class Battle {
   public postBattleLoot: PokemonHeldItemModifier[] = [];
   public escapeAttempts: number = 0;
-  public lastMoveId: MoveId;
+  /** Now holding the {@linkcode Move} instead of just a {@linkcode MoveId}. */
```
is this doc really useful?

flx-sta replied:
Not really I guess

MokaStitcher commented on src/data/ab-attrs/post-dancing-move-ab-attr.ts:
```diff
@@ -33,11 +32,11 @@ export class PostDancingMoveAbAttr extends PostMoveUsedAbAttr {
         if (move.getMove().isSelfStatusMove()) {
           // If the move is a SelfStatusMove (ie. Swords Dance), the Dancer should replicate it on itself
-          globalScene.unshiftPhase(new MovePhase(pokemon, [pokemon.getBattlerIndex()], move, true, true));
+          globalScene.useMove({ pokemon, targets: [pokemon.getBattlerIndex()], move, followUp: true, when: "eager" });
```
```suggestion
          globalScene.useMove({ pokemon, targets: [pokemon.getBattlerIndex()], move, followUp: true, ignorePP: true, when: "eager" });
```
If I understand correctly the second true before was for `ignorePP`?

flx-sta replied:
True. I'm surprised how much I missed here.

https://github.com/Despair-Games/poketernity/blob/d507828e4c4d72dda03ac6cd4178e5d86c1ceade/src/data/ab-attrs/post-dancing-move-ab-attr.ts#L36

flx-sta replied:
It's `ignorePp` though (not `*PP`)

MokaStitcher replied:
All good, it's a lot of changes. Finding this stuff makes me feel better about spending so much time reviewing too 😂

MokaStitcher commented on src/data/ab-attrs/post-dancing-move-ab-attr.ts:
```diff
@@ -33,11 +32,11 @@ export class PostDancingMoveAbAttr extends PostMoveUsedAbAttr {
           // Otherwise, the Dancer must replicate the move on the source of the Dance
           const target = this.getTarget(pokemon, source, targets);
-          globalScene.unshiftPhase(new MovePhase(pokemon, target, move, true, true));
+          globalScene.useMove({ pokemon, targets: target, move, followUp: true, when: "eager" });
```
```suggestion
          globalScene.useMove({ pokemon, targets: target, move, followUp: true, ignorePP: true, when: "eager" });
```
same

MokaStitcher commented on src/battle-scene.ts:
```diff
@@ -3628,4 +3680,130 @@ export default class BattleScene extends SceneBase {
+    const movePhase = new MovePhase(pokemon, targets, move, followUp, ignorePp);
+
+    if ((when === "before" || when === "after") && !phaseId) {
+      throw new Error("phaseId is required for useMove.when === 'before'");
```
```suggestion
      throw new Error("phaseId is required for useMove.when === 'before' or 'after'");
```

MokaStitcher commented on src/data/move-attrs/resist-last-move-type-attr.ts:
```diff
@@ -33,8 +32,8 @@ export class ResistLastMoveTypeAttr extends MoveEffectAttr {
-    const moveData = allMoves[targetMove.moveId];
-    if (moveData.type === ElementalType.STELLAR || moveData.type === ElementalType.UNKNOWN) {
+    const moveData = targetMove.move;
+    if (!moveData || [ElementalType.STELLAR, ElementalType.UNKNOWN].includes(moveData?.type)) {
```
```suggestion
    if (!moveData || [ElementalType.STELLAR, ElementalType.UNKNOWN].includes(moveData.type)) {
```

MokaStitcher commented on src/data/move.ts:
```diff
@@ -886,9 +886,6 @@ export class AttackMove extends Move {
      * {@link https://bulbapedia.bulbagarden.net/wiki/Freeze_(status_condition)}
      * > All damaging Fire-type moves can now thaw a frozen target, regardless of whether or not they have a chance to burn;
      */
-    if (this.type === ElementalType.FIRE) {
```
I'm not sure yet where this was moved, but if the removal is intentional the doc should be removed as well (and probably moved to the spot where this is handled now)

flx-sta replied:
It's been moved into `init-moves`;

I've created a helper function for it that has the comment attached

MokaStitcher commented on src/field/pokemon-move.ts:
```diff
@@ -5,17 +5,17 @@ import { toDmgValue } from "#app/utils";
  * These are the moves assigned to a {@linkcode Pokemon} object.
- * It links to {@linkcode Move} class via the move ID.
- * Compared to {@linkcode Move}, this class also tracks if a move has received.
+ * It links to {@linkcode MoveId} class via the move ID.
```
```suggestion
 * It links to {@linkcode Move} class via the move ID.
```

MokaStitcher commented on src/field/pokemon-move.ts:
```diff
@@ -5,17 +5,17 @@ import { toDmgValue } from "#app/utils";
 
 /**
- * Wrapper class for the {@linkcode Move} class for Pokemon to interact with.
+ * Wrapper class for the {@linkcode MoveId} class for Pokemon to interact with.
```
```suggestion
 * Wrapper class for the {@linkcode Move} class for Pokemon to interact with.
```

MokaStitcher commented on src/field/pokemon.ts:
```diff
@@ -5979,26 +5955,6 @@ export interface AttackMoveResult {
-}
-
 /** Container for Pokemon-specific data that resets at the end of each wave. */
 export class PokemonBattleData {
```
i know this is not really the scope of this PR but for some sort of consistency it might be good to put the other Battle/SummonData classes below in their own file as well

flx-sta replied:
Yeah. There is a lot of stuff like that that needs to happen.
As stated not in the scope of this PR though. Moving files around can be a slippery slope in regards to other PRs

MokaStitcher commented on src/phase.ts:
```diff
@@ -1,10 +1,18 @@
+import { PhaseId } from "#enums/phase-id";
 
 export class Phase {
+  protected _id: PhaseId = PhaseId.UNSPECIFIED;
```
now that we have a phase id, do we still want to keep the isXXXPhase pattern?

Also, would it be better to add it as a parameter to the constructor, to ensure that children classes define their id?

flx-sta replied:
No. If you put it into the contructor it means anyone calling `new <Phase>()` anywhere in the code has to provide the ID which shouldn't be necessary.

flx-sta replied:
> now that we have a phase id, do we still want to keep the isXXXPhase pattern?

Yes. I was too lazy but It's probably better to do that

MokaStitcher replied:
> No. If you put it into the contructor it means anyone calling new <Phase>() anywhere in the code has to provide the ID which shouldn't be necessary.

I meant in the base class constructor, sorry, so all children classes would call super({their id})
but it's not that important I suppose

flx-sta replied:
I forgot that’s possible. That’s a good idea.

why is phase no abstract btw ? Is there a reason ?

MokaStitcher replied:
no clue 😆

MokaStitcher commented on src/phases/battle-end-phase.ts:
```diff
@@ -39,8 +40,7 @@ export class BattleEndPhase extends BattlePhase {
     if (gameMode.isEndless && currentBattle.waveIndex >= 5850) {
-      globalScene.clearPhaseQueue();
-      globalScene.unshiftPhase(new GameOverPhase(true));
+      globalScene.gameOver({ isVictory: true });
```
```suggestion
      globalScene.gameOver({ clearPhaseQueue: true, isVictory: true });
```

MokaStitcher commented on src/phases/title-phase.ts:
```diff
@@ -7,28 +7,32 @@ import { GameMode, getGameMode } from "#app/game-mode";
+import { PhaseId } from "#enums/phase-id";
 
 export class TitlePhase extends Phase {
+  constructor() {
```
weird that it is above the attributes

flx-sta replied:
whops...

MokaStitcher commented on src/utils/encounter-anim-utils.ts:
```diff
@@ -0,0 +1,11 @@
+ * MUST be called after {@linkcode initEncounterAnims()} to load all required animations properly
+ * @param startLoad
+ */
+export async function loadEncounterAnimAssets(startLoad?: boolean): Promise<void> {
```
is there a reason not to put this in anim-utils with loadAnimAssets?

flx-sta replied:
Previously this was causing more depdency violations but not anymore thus I moved them there. Thanks

MokaStitcher commented on src/utils/common-anim-utils.ts:
```diff
@@ -0,0 +1,8 @@
+import { commonAnims } from "#app/data/common-anims";
+import { loadAnimAssets } from "./anim-utils";
+
+export function loadCommonAnimAssets(startLoad?: boolean): Promise<void> {
```
same question

MokaStitcher reviewed (commented):
only challenges and modifier stuff left... will be for later/tomorrow

flx-sta committed:
apply review suggestions

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

MokaStitcher commented on src/utils/anim-utils.ts:
```diff
@@ -38,3 +46,26 @@ export function loadAnimAssets(anims: AnimConfig[], startLoad?: boolean): Promis
 }
+
+/**
+ * Lodas common animation assets to scene.
```
```suggestion
 * Loads common animation assets to scene.
```

DayKev commented on .github/workflows/dependencies.yml:
```diff
@@ -0,0 +1,37 @@
+
+on:
+  # Trigger the workflow on push or pull request,
+  # but only for the main branch
```
```suggestion
  # Trigger the workflow on push or pull request
```

PigeonBar commented on src/utils/arena-tag-type-utils.ts:
```diff
@@ -10,3 +10,16 @@ export const ArenaTrapArenaTagTypes = [
+  ArenaTagType.LIGHT_SCREEN,
+];
+
+export const ConditionalProtectArenaTagTypes = [
```
This hardcoded list can now be replaced with `ConditionalProtectArenaTagTypes`:
https://github.com/Despair-Games/poketernity/blob/962e25cd852f398ee303ad8a8fc19c102aaa2808/src/data/init-moves.ts#L1578-L1580

PigeonBar commented on src/utils/arena-tag-type-utils.ts:
```diff
@@ -10,3 +10,16 @@ export const ArenaTrapArenaTagTypes = [
   ArenaTagType.IMPRISON,
 ];
+
+export const WeakenMoveScreenArenaTagTypes = [
```
This hardcoded list can now be replaced with `WeakenMoveScreenArenaTagTypes`:
https://github.com/Despair-Games/poketernity/blob/962e25cd852f398ee303ad8a8fc19c102aaa2808/src/data/move-attrs/remove-screens-attr.ts#L14-L16

flx-sta replied:
noice

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -3594,26 +3566,22 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-      : this.summonData.tags.find((t) => t.tagType === tagType);
+    const result = this.summonData.tags.find((t) => tagTypes.includes(t.tagType));
+
+    return result ? (result as T) : (result as nil);
```
```suggestion
    return this.summonData.tags.find((t) => tagTypes.includes(t.tagType)) as T | nil;
```

PigeonBar commented on src/utils/battler-tag-type-utils.ts:
```diff
@@ -0,0 +1,38 @@
+import { BattlerTagType } from "#enums/battler-tag-type";
+
+export const SemiInvulnerableBattlerTagTypes = [
```
These hardcoded lists can be replaced by `SemiInvulnerableBattlerTagTypes`:
https://github.com/Despair-Games/poketernity/blob/962e25cd852f398ee303ad8a8fc19c102aaa2808/src/field/pokemon.ts#L4143-L4148
https://github.com/Despair-Games/poketernity/blob/962e25cd852f398ee303ad8a8fc19c102aaa2808/src/data/ab-attrs/post-dancing-move-ab-attr.ts#L21-L26

PigeonBar commented on src/utils/battler-tag-type-utils.ts:
```diff
@@ -0,0 +1,38 @@
+  BattlerTagType.BIND,
+  BattlerTagType.WRAP,
+  BattlerTagType.CLAMP,
+  BattlerTagType.SAND_TOMB,
```
```suggestion
  BattlerTagType.SAND_TOMB,
  BattlerTagType.G_MAX_SAND_TOMB,
```

PigeonBar commented on src/utils/battler-tag-type-utils.ts:
```diff
@@ -0,0 +1,38 @@
+  ...DamagingTrappedBattlerTagTypes,
+];
+
+export const GulpMissleBattlerTagTypes = [BattlerTagType.GULP_MISSILE_ARROKUDA, BattlerTagType.GULP_MISSILE_PIKACHU];
```
```suggestion
export const GulpMissileBattlerTagTypes = [BattlerTagType.GULP_MISSILE_ARROKUDA, BattlerTagType.GULP_MISSILE_PIKACHU];
```

PigeonBar replied:
This hardcoded list can be updated to use `GulpMissileBattlerTagTypes`:
https://github.com/Despair-Games/poketernity/blob/962e25cd852f398ee303ad8a8fc19c102aaa2808/src/data/battler-tags.ts#L2682-L2685

flx-sta committed:
Update src/utils/anim-utils.ts

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
remove `Phase.get*Phase()` methods

instead we are now using Phase.is<*Phase>(PhaseId.*)

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

PigeonBar commented on src/phases/post-summon-phase.ts:
```diff
@@ -22,19 +23,19 @@ export class PostSummonPhase extends PokemonPhase {
     if (
       globalScene.currentBattle.isBattleMysteryEncounter()
-      && pokemon.findTags((t) => t instanceof MysteryEncounterPostSummonTag).length > 0
+      && pokemon.findTags((t) => t.isMysteryEncounterPostSummonTag()).length > 0
```
I think this check and the `is...Tag()` function are redundant, since `lapseTag()` does nothing if the tag doesn't exist.

flx-sta replied:
I'd prefer not to experiment with that in my branch to keep the already large scope small. If you agree please resolve

flx-sta committed:
chore: clean up select-modifier-phase

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -3594,26 +3567,22 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    return this.summonData.tags.find((t) => tagFilter(t));
+    const result = this.summonData.tags.find((t) => tagFilter(t));
+
+    return result ? (result as T) : (result as nil);
```
```suggestion
    return this.summonData.tags.find((t) => tagFilter(t)) as T | nil;
```

flx-sta committed:
set `api.isConnected` default to `true`

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
add `api.ping()` to reload.test beforeEach

revert set `api.isConnected` default to `true`

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
fix errors after beta-merge

MokaStitcher commented on src/phases/encounter-phase.ts:
```diff
@@ -64,11 +64,13 @@ import { PhaseId } from "#enums/phase-id";
  * @extends BattlePhase
  */
 export class EncounterPhase extends BattlePhase {
+  /** @override Can't be just {@linkcode PhaseId.ENCOUNTER} as {@linkcode EncounterPhase} is extended by other phases */
```
confused what this means exactly?

flx-sta replied:
Technically it should be 
```
  override readonly id: PhaseId.ENCOUNTER = PhaseId.ENCOUNTER;
```

So when someone uses 
```
phase.is<EncounterPhase>(PhaseId.MOVE)
``` 
The compiler will complain because only `PhaseId.ENCOUNTER` is valid.

I can't set that though, since an `override` cannot completely ignore the previous type. It can only _zone in_ on the existing one. So if I were to use `id: PhaseId.ENCOUNTER` all the phases that extend `EncounterPhase` would be screaming at me thanks to the compiler not being happy with me overwriting the type with a completely differnt one

MokaStitcher replied:
Encouter (and all other similar comments): 
https://github.com/Despair-Games/poketernity/blob/1b7f747584296fad82ddec71bc4d96a222333401/src/phases/encounter-phase.ts#L67-L68

> "blabla can't be just X" (PhaseId.ENCOUNTER)
> Set value to X (PhaseId.ENCOUNTER)

(which is why I'm confused and don't understand the comment's meaning)


SwitchSummon:
https://github.com/Despair-Games/poketernity/blob/1b7f747584296fad82ddec71bc4d96a222333401/src/phases/switch-summon-phase.ts#L19-L20

> "blabla can't be just X" (PhaseId.SWITCH)
> Set value to Y (PhaseId.SWITCH_SUMMON)

MokaStitcher replied:
Sorry, this was meant as a reply to the switch summon comment

For this one, so the thing that can't be "PhaseId.ENCOUNTER" is the typing of the id, is that it?

flx-sta replied:
So you want them to be unisono

flx-sta replied:
Actually checking again there was some big inconsistencies here and there. Fixed

MokaStitcher commented on src/phases/move-phase.ts:
```diff
@@ -64,6 +66,26 @@ export class MovePhase extends BattlePhase {
   protected cancelled: boolean = false;
 
+  /**
+   * @param followUp Indicates that the move being uses is a "follow-up" - for example, a move being used by Metronome or Dancer.
```
```suggestion
   * @param followUp Indicates that the move being used is a "follow-up" - for example, a move being used by Metronome or Dancer.
```

MokaStitcher commented on src/phases/switch-summon-phase.ts:
```diff
@@ -16,6 +16,9 @@ import { AbAttrFlag } from "#enums/ab-attr-flag";
 import { PhaseId } from "#enums/phase-id";
 
 export class SwitchSummonPhase extends SummonPhase {
+  /** @override Can't be just {@linkcode PhaseId.SWITCH} as {@linkcode SwitchSummonPhase} is extended by other phases. */
```
is this correct? Doesn't follow the same pattern as other similar comments

flx-sta replied:
I'm not sure I follow. The comment follows the same pattern as `EncounterPhase`

MokaStitcher commented on src/modifier/init-modifier-types.ts:
```diff
@@ -0,0 +1,545 @@
+    new PokemonAllMovePpRestoreModifierType("modifierType:ModifierType.MAX_ELIXIR", "max_elixir", -1);
+
+  modifierTypes.PP_UP = () => new PokemonPpUpModifierType("modifierType:ModifierType.PP_UP", "pp_up", 1);
+  (modifierTypes.PP_MAX = () => new PokemonPpUpModifierType("modifierType:ModifierType.PP_MAX", "pp_max", 3)),
```
```suggestion
  modifierTypes.PP_MAX = () => new PokemonPpUpModifierType("modifierType:ModifierType.PP_MAX", "pp_max", 3);
```

MokaStitcher commented on src/modifier/init-modifier-types.ts:
```diff
@@ -0,0 +1,545 @@
+    modifierTypes.SUPER_REPEL = () => new DoubleBattleChanceBoosterModifierType('Super Repel', 10);
+    modifierTypes.MAX_REPEL = () => new DoubleBattleChanceBoosterModifierType('Max Repel', 25),*/
+
+    (modifierTypes.LURE = () =>
+      new DoubleBattleChanceBoosterModifierType("modifierType:ModifierType.LURE", "lure", 10));
```
```suggestion
  modifierTypes.LURE = () =>
    new DoubleBattleChanceBoosterModifierType("modifierType:ModifierType.LURE", "lure", 10);
```

MokaStitcher commented on src/modifier/init-modifier-types.ts:
```diff
@@ -0,0 +1,545 @@
+
+  modifierTypes.PP_UP = () => new PokemonPpUpModifierType("modifierType:ModifierType.PP_UP", "pp_up", 1);
+  (modifierTypes.PP_MAX = () => new PokemonPpUpModifierType("modifierType:ModifierType.PP_MAX", "pp_max", 3)),
+    /*modifierTypes.REPEL = () => new DoubleBattleChanceBoosterModifierType('Repel', 5);
+    modifierTypes.SUPER_REPEL = () => new DoubleBattleChanceBoosterModifierType('Super Repel', 10);
+    modifierTypes.MAX_REPEL = () => new DoubleBattleChanceBoosterModifierType('Max Repel', 25),*/
```
```suggestion

  /*modifierTypes.REPEL = () => new DoubleBattleChanceBoosterModifierType('Repel', 5);
  modifierTypes.SUPER_REPEL = () => new DoubleBattleChanceBoosterModifierType('Super Repel', 10);
  modifierTypes.MAX_REPEL = () => new DoubleBattleChanceBoosterModifierType('Max Repel', 25);*/
```

MokaStitcher commented on src/modifier/init-modifier-pools.ts:
```diff
@@ -0,0 +1,723 @@
+    m.setTier(ModifierTier.COMMON);
+    return m;
+  });
+  modifierPool[ModifierTier.GREAT] = [
```
```suggestion

  modifierPool[ModifierTier.GREAT] = [
```

MokaStitcher commented on src/modifier/init-modifier-pools.ts:
```diff
@@ -0,0 +1,723 @@
+    m.setTier(ModifierTier.GREAT);
+    return m;
+  });
+  modifierPool[ModifierTier.ULTRA] = [
```
```suggestion

  modifierPool[ModifierTier.ULTRA] = [
```

MokaStitcher commented on src/modifier/init-modifier-pools.ts:
```diff
@@ -0,0 +1,723 @@
+    m.setTier(ModifierTier.ULTRA);
+    return m;
+  });
+  modifierPool[ModifierTier.EPIC] = [
```
```suggestion

  modifierPool[ModifierTier.EPIC] = [
```

MokaStitcher commented on src/modifier/init-modifier-pools.ts:
```diff
@@ -0,0 +1,723 @@
+    m.setTier(ModifierTier.EPIC);
+    return m;
+  });
+  modifierPool[ModifierTier.MASTER] = [
```
```suggestion

  modifierPool[ModifierTier.MASTER] = [
```

MokaStitcher reviewed (commented):
I think I'm done???

DayKev reviewed (commented):
Still need to finish reviewing, just submitting this before I forget.

flx-sta committed:
apply review suggestions

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
apply review suggestions

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

PigeonBar commented on src/data/ab-attrs/block-recoil-damage-ab-attr.ts:
```diff
@@ -3,8 +3,14 @@ import { getPokemonNameWithAffix } from "#app/messages";
 import { AbAttr } from "./ab-attr";
+import { AbAttrFlag } from "#enums/ab-attr-flag";
 
 export class BlockRecoilDamageAttr extends AbAttr {
```
```suggestion
export class BlockRecoilDamageAbAttr extends AbAttr {
```

flx-sta committed:
fix inconsistencies for phase id coments & types

MokaStitcher commented on src/phases/common-anim-phase.ts:
```diff
@@ -1,14 +1,18 @@
  * @extends PokemonPhase
  */
 export class CommonAnimPhase extends PokemonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode CommonAnimPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode CommonAnimPhase} is extended by other phases */
```

Much clearer now thank you. "because as" is too many justifications, putting since instead but "because" or "as" would be fine too, just not both :p

MokaStitcher commented on src/phases/encounter-phase.ts:
```diff
@@ -63,6 +64,9 @@ import { MysteryEncounterPhase } from "./mystery-encounter-phases/mystery-encoun
  * @extends BattlePhase
  */
 export class EncounterPhase extends BattlePhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode EncounterPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode EncounterPhase} is extended by other phases */
```

MokaStitcher commented on src/phases/hit-check-phase.ts:
```diff
@@ -18,17 +17,26 @@ import { MoveFlags } from "#enums/move-flags";
  * @see {@linkcode hitCheck}
  */
 export abstract class HitCheckPhase extends PokemonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode HitCheckPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode HitCheckPhase} is extended by other phases */
```

MokaStitcher commented on src/phases/modifier-reward-phase.ts:
```diff
@@ -1,10 +1,14 @@
 import i18next from "i18next";
 
 export class ModifierRewardPhase extends Phase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode ModifierRewardPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode ModifierRewardPhase} is extended by other phases */
```

MokaStitcher commented on src/phases/next-encounter-phase.ts:
```diff
@@ -1,11 +1,15 @@
  * @extends EncounterPhase
  */
 export class NextEncounterPhase extends EncounterPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode NextEncounterPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode NextEncounterPhase} is extended by other phases */
```

MokaStitcher commented on src/phases/summon-phase.ts:
```diff
@@ -10,12 +10,15 @@ import { getPokemonNameWithAffix } from "#app/messages";
+import { PhaseId } from "#enums/phase-id";
 
 export class SummonPhase extends PartyMemberPokemonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode SummonPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode SummonPhase} is extended by other phases */
```

MokaStitcher commented on src/phases/switch-summon-phase.ts:
```diff
@@ -1,19 +1,24 @@
+import { PhaseId } from "#enums/phase-id";
 
 export class SwitchSummonPhase extends SummonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} because as {@linkcode SummonPhase} is extended by other phases */
```
```suggestion
  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode SummonPhase} is extended by other phases */
```

MokaStitcher reviewed (dismissed):
👀

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -5958,6 +5934,7 @@ interface AbilityData {
 
 export interface TurnMove {
   moveId: MoveId;
+  move: Move | undefined;
```
```suggestion
  move?: Move;
```

flx-sta replied:
No, I did this on purpose. I want people to consciously set this to `undefined` so they don't accidentially skip it.

In the long run I'd like to get rid of `moveId` in general

DayKev replied:
Maybe it should be `null` instead then, since we're using this as an intentional "no value".

DayKev replied:
```suggestion
  // TODO: change to `| null`
  move: Move | undefined;
```

DayKev replied:
(Last nit then 👍 )

flx-sta committed:
comments: replace `because as` with `since`

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

PigeonBar commented on src/utils/move-utils.ts:
```diff
@@ -0,0 +1,84 @@
+  ...params: Parameters<TAttr["apply"]>
+): void {
+  const [user, target, move, ...args] = params;
+  if (move.isChargingAttackMove() || move.isChargingSelfStatusMove()) {
```
```suggestion
  if (move.isChargingMove()) {
```
These two extra "isCharging...Move()" functions should no longer be needed

PigeonBar commented on src/data/init-moves.ts:
```diff
@@ -0,0 +1,4085 @@
+ * @source {@link https://bulbapedia.bulbagarden.net/wiki/Freeze_(status_condition) | Bulbapedia - Freeze (Status Condition)}
+ */
+function addFireMovesThawFrozenTargetAttribute(move: Move) {
+  if (move.type === ElementalType.FIRE) {
```
```suggestion
  if (move.type === ElementalType.FIRE && move.isAttackMove()) {
```
I just realized that this logic was originally in the constructor for `AttackMove`, not `Move`

PigeonBar commented on src/plugins/api/api.ts:
```diff
@@ -77,6 +94,16 @@ export class Api extends ApiBase {
+   * Pings the server (via {@linkcode getGameTitleStats}) and updates {@linkcode _isConnected} accordingly.
+   */
+  async ping() {
+    if (this.isLocal) {
+      const titleStats = await this.getGameTitleStats();
+      this._isConnected = !!titleStats;
+      console.log("isLocalServerConnected:", this.isConnected);
+    }
```
```suggestion
    const titleStats = await this.getGameTitleStats();
    this._isConnected = !!titleStats;
    console.log("isLocalServerConnected:", this.isConnected);
```
(Since `_isConnected` has no default value, we also need to update its value when `isLocal` is false)

flx-sta committed:
apply review suggestions

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

MokaStitcher commented on src/field/pokemon.ts:
```diff
@@ -3594,26 +3567,15 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
-  findTag(tagFilter: (tag: BattlerTag) => boolean) {
-    if (!this.summonData) {
```
Don't think this check should be removed

flx-sta replied:
Generally this makes me wonder why `summonData` isn't marked as `?` or ` | null`

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta committed:
restore `summonData` check for findTag

DayKev commented on src/battle-scene.ts:
```diff
@@ -195,6 +202,52 @@ export interface InfoToggle {
+
+const DEBUG_RNG = false;
+
+export const startingWave = Overrides.STARTING_WAVE_OVERRIDE || 1;
```
```suggestion
const startingWave = Overrides.STARTING_WAVE_OVERRIDE || 1;
```
Unless you changed it, this isn't used in any other files.

flx-sta replied:
I was actually thinking of moving that into `constants.ts`. Does this make sense?

Same for `DEBUG_RNG` which should then be controlled through env variables

DayKev replied:
Well if it's only going to be used in this file then there's no need to move it. Expanding the use of these would probably be a different PR though.

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

MokaStitcher reviewed (dismissed):
Approving, but let's wait for the final DayKev's review before merging

flx-sta committed:
Update src/battle-scene.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

DayKev commented on src/data/apply-ab-attrs.ts:
```diff
@@ -76,9 +104,11 @@ function applyAbAttrsInternal<TAttr extends AbAttr>(
 
-export function applyRevealedAbAttrs<TAttr extends AbAttr>(
-  attrType: AbstractConstructor<TAttr>,
+function applyRevealedAbAttrs<TAttr extends AbAttr>(
```
This seems like it should be a parameter instead of an entire function. 🤔
Out of scope for this PR probably, but something to remember for later.

flx-sta replied:
Created a draft task on the project board:
https://github.com/orgs/Despair-Games/projects/3/views/3?pane=issue&itemId=97010293

DayKev commented on src/data/battle-anims/move-anim.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  protected override isHideUser(): boolean {
+    return allMoves[this.moveId].hasFlag(MoveFlags.HIDE_USER);
```
Another refactor that needs to happen: replace `.hasFlag()` with `.checkFlag()` throughout the codebase.

flx-sta replied:
Created task item on the board:
https://github.com/orgs/Despair-Games/projects/3/views/3?pane=issue&itemId=97010472

DayKev commented on src/battle-scene.ts:
```diff
@@ -3626,4 +3679,130 @@ export default class BattleScene extends SceneBase {
     return encounter;
   }
+
+  nextBattle(isVictory: boolean): void {
```
Note to self: I think there's places in the codebase where a pushed `BattleEndPhase` isn't immediately followed by a pushed `NewBattlePhase`, keep an eye out for this.

flx-sta replied:
There is a couple:

### Victory Phase

https://github.com/Despair-Games/poketernity/blob/b15e02d40a91c83b025bd9b6793dec9e4daf74c0/src/phases/victory-phase.ts#L57

---

### Encounter Phase **Utils**

https://github.com/Despair-Games/poketernity/blob/b15e02d40a91c83b025bd9b6793dec9e4daf74c0/src/data/mystery-encounters/utils/encounter-phase-utils.ts#L879

https://github.com/Despair-Games/poketernity/blob/b15e02d40a91c83b025bd9b6793dec9e4daf74c0/src/data/mystery-encounters/utils/encounter-phase-utils.ts#L846

DayKev commented on src/data/move-conditions/fail-if-last-condition.ts:
```diff
@@ -0,0 +1,9 @@
+import { PhaseId } from "#enums/phase-id";
+
+export const failIfLastCondition: MoveConditionFunc = (_user: Pokemon, _target: Pokemon, _move: Move) =>
+  globalScene.phaseQueue.find((phase) => phase.is<MovePhase>(PhaseId.MOVE)) !== undefined;
```
Interesting that this doesn't use the `BattleScene.findPhase()` function.

flx-sta replied:
Added a new `hasPhase()` method

DayKev commented on src/data/move-conditions/last-move-copiable-condition.ts:
Note to self: this is deleted in the Metronome/etc refactor.

DayKev reviewed (commented):
![image](https://github.com/user-attachments/assets/c46a186e-7afe-42cf-ac02-2b632c8c0b5d)
Gonna do the other half "tomorrow".

flx-sta committed:
add `hasFlag` method to BattleScene

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

DayKev commented on src/data/species-form-change-triggers/meloetta-form-change-post-move-trigger.ts:
```diff
@@ -0,0 +1,19 @@
+
+export class MeloettaFormChangePostMoveTrigger extends SpeciesFormChangePostMoveTrigger {
+  override canChange(pokemon: Pokemon): boolean {
+    if (globalScene.gameMode.hasChallenge(Challenges.SINGLE_TYPE)) {
```
```suggestion
    // TODO: improve this (should only block the form change in Psychic or Fighting mono-type, not Normal)
    if (globalScene.gameMode.hasChallenge(Challenges.SINGLE_TYPE)) {
```
Something I forgot about when implementing this.

DayKev commented on src/field/pokemon.ts:
```diff
@@ -1672,24 +1641,24 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
    * Accounts for all the various effects which can affect whether an ability will be present or
    * in effect, and both passive and non-passive.
-   * @param attrType - {@linkcode AbAttr} The ability attribute to check for.
+   * @param abAttrFlag - The {@linkcode AbAttrFlag} to check for in the abilities
```
The wording for this is kinda awkward, but I'm not sure what to change it to.

flx-sta replied:
done

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3566,13 +3537,15 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
 
     if (!cancelled.value && newTag.canAdd(this)) {
-      this.summonData.tags.push(newTag);
+      this.summonData?.tags.push(newTag);
```
If `summonData` isn't initialized then `BattlerTag`s can't be applied to the pokemon? That seems like an issue.

flx-sta replied:
Well tbh. This change wasn't necessary but technically `summonData` would have to be marked as `?` since it CAN be `undefined` apparently.
See https://github.com/Despair-Games/poketernity/blob/42b80784744a3c2ce3687ff7a0ba18e28be027dc/src/field/pokemon.ts#L3563-L3565

I'll just undo this change for now and we should fix this some other time

DayKev replied:
Yeah this seems like something that needs to be looked at in the future, `summonData` probably shouldn't be able to be `undefined`.

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3581,26 +3554,18 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    return tagType instanceof Function
-      ? this.summonData.tags.find((t) => t instanceof tagType)
-      : this.summonData.tags.find((t) => t.tagType === tagType);
+    return this.summonData.tags.find((t) => tagTypes.includes(t.tagType)) as T | nil;
```
```suggestion
    return this.summonData.tags.find((t) => tagTypes.includes(t.tagType)) as T | undefined;
```
It might make sense to have the `summonData` check also return `undefined`, dunno if it's useful for `getTag()` to indicate whether `summonData` is defined or not.

flx-sta replied:
I've adopted this fromt he original code:
https://github.com/Despair-Games/poketernity/blob/67d278a1b60c8d878404c72948950559076811fe/src/field/pokemon.ts#L3589C1-L3593C74

I'd rather not chnage it in the scope fo this PR

DayKev replied:
The original code returned `undefined` here; I was talking about potentially replacing the `null` return that's above it with `undefined` instead since when would it be useful to know why `getTag` didn't return the `BattlerTag` you're checking for.

flx-sta replied:
@DayKev Not sure I follow...

DayKev replied:
https://github.com/Despair-Games/poketernity/blob/67d278a1b60c8d878404c72948950559076811fe/src/field/pokemon.ts#L3595-L3602
Originally, `getTag()` would return `null` if `pokemon.summonData` was `undefined`. If that's not the case, then `tags.find()` would return either a matching `BattlerTag`, or `undefined` if there was none. Therefore if `getTag()` returns `null` then you know that the only reason it did so was because `pokemon.summonData` was `undefined`, and if `getTag()` returned `undefined` then you know the only reason it did so is because the `BattlerTag`(s) you searched for weren't on the pokemon.

Following on from that, there are two things:
- First, I was thinking that there is probably no reason for there to be that distinction, because when calling `getTag()` you are never going to care *why* the `BattlerTag` you searched for wasn't on the pokemon, just whether it was there or not.
- Second, you changed the return type of the `tags.find()` code path from `T | undefined` (which matches what `tags.find()` will output) to `T | nil`, and my code suggestion was to return it back to `T | undefined` to match previous behavior.

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3581,26 +3554,18 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
       return null;
     }
-    return this.summonData.tags.find((t) => tagFilter(t));
+    return this.summonData.tags.find((t) => tagFilter(t)) as T | nil;
```
```suggestion
    return this.summonData.tags.find((t) => tagFilter(t)) as T | undefined;
```
Same for this function.

flx-sta replied:
https://github.com/Despair-Games/poketernity/blob/67d278a1b60c8d878404c72948950559076811fe/src/field/pokemon.ts#L3589C1-L3593C74

Adopted from here

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3581,26 +3554,18 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-  getTag<T extends BattlerTag>(tagType: AbstractConstructor<T>): T | nil;
-
-  getTag(tagType: BattlerTagType | AbstractConstructor<BattlerTag>): BattlerTag | nil {
+  getTag<T extends BattlerTag>(...tagTypes: BattlerTagType[]): T | nil {
```
```suggestion
  getTag<T extends BattlerTag = BattlerTag>(...tagTypes: BattlerTagType[]): T | nil {
```
Is this intentionally different from `findTag()`?

flx-sta replied:
negative

DayKev commented on src/utils/party-utils.ts:
Maybe it should be called `party-ui-utils.ts`?

flx-sta replied:
I thought there mightbe more "party" related stuff coming in, but sure

DayKev commented on src/field/pokemon.ts:
```diff
@@ -5945,6 +5909,7 @@ interface AbilityData {
 
 export interface TurnMove {
   moveId: MoveId;
+  move: Move | undefined;
```
Having both the `Move` and the `MoveId` seems redundant.

flx-sta replied:
It is. I'll see if i can get rid of it here actually. I tried to not touch too much extra but this turned out to all have to be changed

flx-sta replied:
It would involve some extra null checks etc. I'd rather add it to the list of TODOs and not increase the scope of the PR again.

flx-sta replied:
I've created a task on the board

DayKev replied:
Yeah that's fine.

DayKev commented on src/phase.ts:
```diff
@@ -1,10 +1,10 @@
+import { PhaseId } from "#enums/phase-id";
 
 export class Phase {
+  /** The identifier of the phase. Unique per phase, but **not* per instance! */
```
```suggestion
  /** The identifier of the phase. Unique per phase, but *not* per instance! */
```

flx-sta replied:
? I wanted this to be bold

DayKev replied:
You have an uneven number of `*`, either it should be 2 on both sides or 1.

flx-sta replied:
Ooooh.. okay thanks

DayKev commented on src/phase.ts:
```diff
@@ -15,19 +15,7 @@ export class Phase {
-
-  isSelectModifierPhase(): this is SelectModifierPhase {
-    return false;
+  is<T extends Phase = Phase>(phaseId: T["id"]): this is T {
```
```suggestion
  public is<T extends Phase = Phase>(phaseId: T["id"]): this is T {
```

DayKev commented on src/phase.ts:
```diff
@@ -1,10 +1,10 @@
-import type { SwitchPhase } from "#app/phases/switch-phase";
+import { PhaseId } from "#enums/phase-id";
 
 export class Phase {
```
I wonder, should this class be `abstract`?

flx-sta replied:
I was wondering the same recently. Even asked on discord if anyone knew why it isn't :D

DayKev replied:
If you don't find any `new Phase()` in the codebase then I think it's safe to make it `abstract`.

flx-sta replied:
I did and it seems to workf ine

DayKev commented on src/phases/common-anim-phase.ts:
```diff
@@ -1,14 +1,18 @@
  * @extends PokemonPhase
  */
 export class CommonAnimPhase extends PokemonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode CommonAnimPhase} is extended by other phases */
```
What do you mean?

flx-sta replied:
Other have `PhaseId.<ENUM_KEY>` but phases that are extended by others can't have that

DayKev commented on src/phases/hit-check-phase.ts:
```diff
@@ -18,17 +17,26 @@ import { MoveFlags } from "#enums/move-flags";
  */
 export abstract class HitCheckPhase extends PokemonPhase {
+  /** @override **Must** use generic {@linkcode PhaseId} since {@linkcode HitCheckPhase} is extended by other phases */
+  override readonly id: PhaseId = PhaseId.HIT_CHECK;
```
Since this is an `abstract` phase does it need an `id`? You didn't add one to the other `abstract` phases.
Actually that makes me realize I never renamed this file to `abstract-hit-check-phase.ts` after it got added.

flx-sta replied:
true...

DayKev commented on src/phases/select-challenge-phase.ts:
```diff
@@ -1,8 +1,15 @@
+  constructor() {
+    super();
+  }
+
```
```suggestion
```

DayKev commented on src/phases/move-effect-phase.ts:
```diff
@@ -56,6 +55,10 @@ export class MoveEffectPhase extends HitCheckPhase {
   /** Is this the last strike of a move? */
   private lastHit: boolean;
 
+  constructor(battlerIndex: BattlerIndex, targets: BattlerIndex[], move: PokemonMove) {
+    super(battlerIndex, targets, move);
+  }
+
```
```suggestion
```
Why did you add a constructor? Since it's the same as the parent class it should be inherited I would think.

flx-sta replied:
It's probably a left-over from the previous ID approach and I forgot to delete it again

DayKev commented on src/plugins/api/api.ts:
```diff
@@ -16,6 +16,11 @@ export class Api extends ApiBase {
   public readonly admin: AdminApi;
   public readonly savedata: SavedataApi;
 
+  /** Wheter the hostname is 'localhost' or an IP address, and ensure a port is specified. */
```
These tsdocs need to be on the getters for them to show up when using `api.isLocal` and `api.isConnected`.

DayKev commented on src/utils/arena-tag-type-utils.ts:
In the future we might want to use `Object.freeze(...)` on all these consts that are supposed to be read-only throughout the codebase.

flx-sta replied:
Totally forgot that a thing. Good point! Could also be another obfuscation for potential "hacking" scripts

DayKev commented on src/utils/item-utils.ts:
```diff
@@ -0,0 +1,15 @@
+import { globalScene } from "#app/global-scene";
+import { getPokemonNameWithAffix } from "#app/messages";
+import type { PokemonHeldItemModifier } from "#app/modifier/modifier";
+import { t } from "i18next";
```
```suggestion
import i18next from "i18next";
```

flx-sta replied:
`t` is perfectly fine

DayKev replied:
It's inconsistent with the rest of the codebase.

flx-sta replied:
Maybe the rest of the code-base in inconsistent with this ... fine........

DayKev replied:
Maybe it's something we can change, but that's a different PR. Though I'm not sure I'm a fan of a single-letter function like that.

DayKev commented on src/utils/item-utils.ts:
```diff
@@ -0,0 +1,15 @@
+    (m) => m.isPokemonHeldItemModifier() && m.pokemonId === pokemon.id && m.matchType(modifier),
+  );
+  if (matchingModifier && matchingModifier.stackCount === matchingModifier.getMaxStackCount()) {
+    return t("partyUiHandler:tooManyItems", { pokemonName: getPokemonNameWithAffix(pokemon) });
```
```suggestion
    return i18next.t("partyUiHandler:tooManyItems", { pokemonName: getPokemonNameWithAffix(pokemon) });
```

flx-sta replied:
I refuse. `t` is perfectly fine and less bloaty

DayKev commented on src/utils/modifier-pool-utils.ts:
```diff
@@ -0,0 +1,31 @@
+import { ModifierPoolType } from "#enums/modifier-pool-type";
+
+export function getModifierPoolForType(poolType: ModifierPoolType): ModifierPool {
+  let pool: ModifierPool;
+  switch (poolType) {
+    case ModifierPoolType.PLAYER:
+      pool = modifierPool;
+      break;
+    case ModifierPoolType.WILD:
+      pool = wildModifierPool;
+      break;
+    case ModifierPoolType.TRAINER:
+      pool = trainerModifierPool;
+      break;
+    case ModifierPoolType.ENEMY_BUFF:
+      pool = enemyBuffModifierPool;
+      break;
+    case ModifierPoolType.DAILY_STARTER:
+      pool = dailyStarterModifierPool;
+      break;
+  }
+  return pool;
```
```suggestion
  switch (poolType) {
    case ModifierPoolType.PLAYER:
      return modifierPool;
    case ModifierPoolType.WILD:
      return wildModifierPool;
    case ModifierPoolType.TRAINER:
      return trainerModifierPool;
    case ModifierPoolType.ENEMY_BUFF:
      return enemyBuffModifierPool;
    case ModifierPoolType.DAILY_STARTER:
      return dailyStarterModifierPool;
  }
```

DayKev commented on test/moves/disable.test.ts:
```diff
@@ -56,7 +57,11 @@ describe("Moves - Disable", () => {
-    expect(playerMon.getMoveHistory()[0]).toMatchObject<TurnMove>({ moveId: MoveId.DISABLE, result: MoveResult.FAIL });
+    expect(playerMon.getMoveHistory()[0]).toMatchObject<TurnMove>({
+      moveId: MoveId.DISABLE,
+      move: expect.any(Move),
```
Uh, I assume this is basically "disabling" this field from being checked by the `expect()`?

flx-sta replied:
It's simply telling it that I expect any class of type `Move`.

https://jestjs.io/docs/expect#expectanyconstructor

> `expect.any(constructor)` matches anything that was created with the given constructor or if it's a primitive that is of the passed type. You can use it inside `toEqual` or toHaveBeenCalledWith instead of a literal value. For example, if you want to check that a mock function is called with a number:
> 
> ```ts
> class Cat {}
> function getCat(fn) {
>   return fn(new Cat());
> }
> 
> test('randocall calls its callback with a class instance', () => {
>   const mock = jest.fn();
>   getCat(mock);
>   expect(mock).toHaveBeenCalledWith(expect.any(Cat));
> });
> 
> function randocall(fn) {
>   return fn(Math.floor(Math.random() * 6 + 1));
> }
> 
> test('randocall calls its callback with a number', () => {
>   const mock = jest.fn();
>   randocall(mock);
>   expect(mock).toHaveBeenCalledWith(expect.any(Number));
> });

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta commented on src/data/init-moves.ts:
```diff
@@ -0,0 +1,4085 @@
+      .ignoresProtect(),
+    new AttackMove(MoveId.G_MAX_RAPID_FLOW, ElementalType.WATER, MoveCategory.PHYSICAL, 10, -1, 10, -1, 0, 8)
+      .gMaxMove()
+      .ignoresProtect(),
```
@Tempo-anon Can you double-check those for me again please? I had to manually merge them from here:
https://github.com/Despair-Games/poketernity/pull/374/files#diff-4c037804d90ef944f62c1e3b70ff7261bf0cd572b40626f8691b471c6711cd38

Also: Is `G MAX Fireball` really `Physical` category?

Tempo-anon replied:
Looks correct to me. Cinderace has 116atk compared to 65spA and its signature move pyro ball which G-Max Fireball is based off of is physical.

flx-sta committed:
apply review suggestions

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
fix init-moves for GMAX

flx-sta committed:
add proper messaghes for `all-moves` test

flx-sta committed:
fix comment in `phase.ts`

flx-sta committed:
...... use `i18next.t` instead of simply, short an easy `t`

PigeonBar commented on src/utils/party-ui-utils copy.ts:
Is this copy file still supposed to be here?

flx-sta replied:
nope. thakns

PigeonBar commented on test/data/all_moves.test.ts:
```diff
@@ -56,18 +57,35 @@ describe("All Moves", async () => {
           const actualHasFlag = pktyMove.hasFlag(flagsToCheck[f]);
           const expectedHasFlag = move.flags.includes(Number(f));
-          expect(actualHasFlag).toBe(expectedHasFlag);
+          expect(actualHasFlag, `${MoveId[pktyMove.id]} is missing flag ${Number(f)}!`).toBe(expectedHasFlag);
```
```suggestion
          const errOutput = `${MoveId[pktyMove.id]}'s usage of flag ${MoveFlags[flagsToCheck[f]]} should be ${expectedHasFlag} but is ${actualHasFlag}!`
          expect(actualHasFlag, errOutput).toBe(expectedHasFlag);
```
(So that the output is also accurate if a move incorrectly has an extra flag)

flx-sta committed:
Merge branch 'beta' into refactor/depcruise-03

flx-sta commented on src/data/init-common-anims.ts:
> [!CAUTION]
> EXTRA ATTENTION TO THIS FILE
> All current/future changes to `all-abilities` have to be manually adopted into this file. Make sure nothing is getting lost!

DayKev replied:
Probably mention it in the PR description so it's more noticeable.

flx-sta replied:
I put it there too, to make sure both routes are covered

flx-sta commented on src/data/init-trainer-type-dialogue.ts:
> [!CAUTION]
> EXTRA ATTENTION TO THIS FILE
> All current/future changes to `all-moves` have to be manually adopted into this file. Make sure nothing is getting lost!

flx-sta committed:
Update test/data/all_moves.test.ts

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
apply review suggestions

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>
Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
chore: es-lint

DayKev commented on src/data/move.ts:
```diff
@@ -805,23 +795,25 @@ export abstract class Move implements Localizable {
     power.value *= typeChangeMovePowerMultiplier.value;
 
-    const typeBoost = source.findTag((t) => t instanceof TypeBoostTag && t.boostedType === this.type) as TypeBoostTag;
+    const typeBoost = source.findTag<TypeBoostTag>((t) => t.isTypeBoostTag() && t.boostedType === this.type);
```
Something to be done in a follow-up, replace the `.is*Tag()` functions with checks of `*TagTypes` arrays.

DayKev replied:
Added an item to the project board.

Tempo-anon reviewed (dismissed):
The G-Max stuff looks correct

flx-sta committed:
remove `moveId` from `TurnMove`

- move TurnMove into it's own file
- move QueuedMove into it's own file
- move AttackMoveResult into it's own file

MokaStitcher commented on src/data/battler-tags.ts:
```diff
@@ -300,14 +300,14 @@ export class DisabledTag extends MoveRestrictionBattlerTag {
     super.onAdd(pokemon);
 
-    const move = pokemon
+    const turnMove = pokemon
```
looks like this could be replaced by a call to `getLastValidMove`?

<sub> maybe getLastValidMove shoud be renamed to getLastValidMoveId too

flx-sta replied:
> maybe getLastValidMove shoud be renamed to getLastValidMoveId too

done

> looks like this could be replaced by a call to getLastValidMove?

This method only exists on `GorillaTacticsTag`

flx-sta replied:
I've moved it up a level and use it now

MokaStitcher reviewed (dismissed):
just a nit

flx-sta committed:
move `getLastValidMove` into `MoveRestrictionBattlerTag`

- change return type of it  to Move (instead of just MoveId)
- Make use of it in DisableTag

flx-sta commented:
Thank you for all the reviews/feedback/comments etc

Tempo-anon reviewed (approved):
safetycat
